src/profile-logic/stack-timing.js

```diff
@@ -57,135 +56,91 @@ export type StackTiming = {|
 
 export type StackTimingByDepth = Array<StackTiming>;
 
-type LastSeen = {
-  startTimeByDepth: number[],
-  callNodeIndexByDepth: IndexIntoCallNodeTable[],
-};
-
 /**
  * Build a StackTimingByDepth table from a given thread.
  */
 export function getStackTimingByDepth(
   samples: SamplesLikeTable,
+  sampleCallNodes: Array<IndexIntoCallNodeTable | null>,
   callNodeInfo: CallNodeInfo,
   maxDepthPlusOne: number,
   interval: Milliseconds
 ): StackTimingByDepth {
-  const { callNodeTable, stackIndexToCallNodeIndex } = callNodeInfo;
+  const { callNodeTable } = callNodeInfo;
+  const {
+    prefix: callNodeTablePrefixColumn,
+    subtreeRangeEnd: callNodeTableSubtreeRangeEndColumn,
+    depth: callNodeTableDepthColumn,
+  } = callNodeTable;
   const stackTimingByDepth = Array.from({ length: maxDepthPlusOne }, () => ({
     start: [],
     end: [],
     callNode: [],
     length: 0,
   }));
 
-  const lastSeen: LastSeen = {
-    startTimeByDepth: [],
-    callNodeIndexByDepth: [],
-  };
-
-  // Go through each sample, and push/pop it on the stack to build up
-  // the stackTimingByDepth.
-  let previousDepth = -1;
-  for (let i = 0; i < samples.length; i++) {
-    const stackIndex = samples.stack[i];
-    const sampleTime = samples.time[i];
-
-    // If this stack index is null (for instance if it was filtered out) then pop back
-    // down to the base stack.
-    if (stackIndex === null) {
-      _popStacks(stackTimingByDepth, lastSeen, -1, previousDepth, sampleTime);
-      previousDepth = -1;
-    } else {
-      const callNodeIndex = stackIndexToCallNodeIndex[stackIndex];
-      const depth = callNodeTable.depth[callNodeIndex];
-
-      // Find the depth of the nearest shared stack.
-      const depthToPop = _findNearestSharedCallNodeDepth(
-        callNodeTable,
-        callNodeIndex,
-        lastSeen,
-        depth
-      );
-      _popStacks(
-        stackTimingByDepth,
-        lastSeen,
-        depthToPop,
-        previousDepth,
-        sampleTime
-      );
-      _pushStacks(callNodeTable, lastSeen, depth, callNodeIndex, sampleTime);
-      previousDepth = depth;
-    }
+  if (samples.length === 0) {
+    return stackTimingByDepth;
   }
 
-  // Pop the remaining stacks
-  const lastIndex = samples.length - 1;
-  const endingTime = samples.time[lastIndex] + interval;
-  _popStacks(stackTimingByDepth, lastSeen, -1, previousDepth, endingTime);
+  let deepestOpenBoxCallNodeIndex = -1;
+  let deepestOpenBoxDepth = -1;
+  const openBoxStartTimeByDepth = new Float32Array(maxDepthPlusOne);
```

Comment:
As a quick note, I find this quite clever that the only way we remember open boxes in this algorithm is from the deepest open depth, then we recover all the info for building the boxes using the deepest call node index information.